Adding offline mode for dreq_content, adding config file

sol1105 opened the pull request:
Adding offline mode for dreq_content
- Modified functions to accept the offline kwarg.

Adding configuration file
- Set up creation and loading of the configuration file.
- Added a decorator to provide configuration key-value pairs as kwargs if not already set in the function call.
- Configured an entry point to initialize/adjust the configuration via the command line in pyproject.toml.

Adjusted tests for dreq_content.py
- Updated / fixed tests
- Moved test_dreq_content.py to /tests

Removed nesting for data_request_api.content

Updated README.md

Some new tests have yet to be added (once the basic concept is approved).

sol1105 committed:
Adding offline mode for dreq_content
- Modified functions to accept the offline kwarg.

Adding configuration file
- Set up creation and loading of the configuration file.
- Added a decorator to provide configuration key-value pairs as kwargs if not already set in the function call.
- Configured an entry point to initialize/adjust the configuration via the command line in pyproject.toml.

Adjusted tests for dreq_content.py
- Updated / fixed tests
- Moved test_dreq_content.py to /tests

Removed nesting for data_request_api.content

Updated README.md

sol1105 committed:
Follow-up on removal of content nesting. Update doc strings for dreq_content.

sol1105 commented:
@matthew-mizielinski The test `test_api_and_html_request` retrieves the list of tags  via a request to the GitHub API (https://api.github.com/repos/CMIP-Data-Request/CMIP7_DReq_content/) and via parsing of the common repository website (https://github.com/CMIP-Data-Request/CMIP7_DReq_content/) to then assert the retrieved list of tags is the same for both options.

Usually the API call is made as follows: `dc._send_api_request(dc.REPO_API_URL, REPO_PAGE_URL, "tags")`. The function tries the API request and if that fails, it falls back to parsing the repo website. For the test I pass an empty string as the REPO_PAGE_URL to make sure the test fails if the API request fails, and does not fall back to parsing the repo website. I guess since you have generally problems with reaching the GitHub API (which is why the fallback option was implemented in the first place ;-) ) the test can only fail for you. And the error message (`Invalid URL 'tags': No scheme supplied. Perhaps you meant https://tags`) makes sense for the fallback being triggered but failing as intended. 

Do you suggest some changes to the test or the function? If an empty string is passed as REPO_PAGE_URL I could the function not have trigger the fallback but fail with the original error message to make clear what happens.

sol1105 committed:
merge main

sol1105 committed:
Address github api request failing via Github actions.

sol1105 commented:
Apparently the github workflows can also not access the GitHub API, so I had the test xfail when the GitHub API  request does not work.

matthew-mizielinski commented on data_request_api/stable/utilities/config.py:
I would be tempted to move this module, or at least the `main()` function under the `command_line` package to make it easier to find the code used for the command line tools.

sol1105 replied:
Yes, that makes sense

matthew-mizielinski commented on data_request_api/stable/content/dreq_content.py:
```diff
@@ -582,12 +652,12 @@ def load(version="latest_stable", **kwargs):
                 return json.load(f)
         else:
-            return ce.map_data(json.load(f), mapping_table)
+            return json.load(f)  # ce.map_data(json.load(f), mapping_table)
```
remove comment?

matthew-mizielinski commented:
@sol1105, before this morning's updates the test failure was down to line 105, `test_dreq_content.test_api_and_html_request`
The change you put into the `test_api_and_html_request` to xfail solves the problem that I have had -- previously the line
```
tags1 = set(dc._send_api_request(dc.REPO_API_URL, "", "tags"))
```
was leading to failure as the empty string in the second argument is used when I ran the code as the api request always fails here at the Met Office (I think this is down to our firewall configuration, but I could be wrong).

The tests all pass for me now

matthew-mizielinski commented:
At present we've got the  data request content going into a directory under `data_request_api/stable/content/dreq_res` -- adding this location into the config file would be a really useful addition (in the installation of my system users won't have write access to the location in which this code is installed). I'm happy for this to be separated into a different issue, which I can take, if it isn't trivial to fix.

Just an observation (no action required): The use of the decorator to propagate the config options around isn't something I've seen before, but does appear to work well.  Another option would be to use a singleton dataclass (example [here](https://stackoverflow.com/a/77864020)), and instantiate this everywhere these arguments were used.

sol1105 committed:
Updated config.py
- Segregated CLI code into separate module
  Available commands are: init, list, reset and <key> <value> to update configuration key with value
- Added 'cache_dir' to default configuration to specify the location of the cached data request content files
- Added _sanity_check to check type and value of key and value pairs of the configuration dictionary
- Added global variable CONFIG at module level holding the loaded/read configuration
- Added and updated tests for config module, config CLI and the decorator supplying functions with the config kwargs
- Added hidden command line argument --cfgfile to enable custom configuration files for pytest tests

sol1105 committed:
Updated dreq_content.py
- Directory with content cache (_dreq_res) configurable via configuration file
- Default location in configuration in user home directory: ~/.CMIP7_data_request_api_cache/
- 'get_cached' will no longer list both, release and raw exports as default, but follow
  the default setting of the configuration file if the export kwarg is not specified
- 'delete' will no longer delete both, release and raw exports as default, but follow
  the default setting of the configuration file if the export kwarg is not specified
- 'delete' will raise a ValueError if an invalid export kwarg is specified
- 'load' will not anymore attempt to consolidate the loaded data request by default.
  The consolidate option is still in development and not fully functioning.
- Added and updated the tests

sol1105 committed:
Updated config CLI entry point in pyproject.toml

sol1105 committed:
Updated config.py: Allow empty configuration files

sol1105 commented:
@matthew-mizielinski 
I updated the cache location to be customizable via the config file. The default is now: `~/.CMIP7_data_request_api_cache/`. However, changing the location does not have the effect of the "old" cache being moved to the new location.
I moved the config CLI part to another module as suggested. Tests were added for the config CLI, config in general and the decorator. Also some tests for the content were updated / added.
When you have time it would be nice if you had another look. At first glance, the failing tests on windows do not have anything to do with the content and config part.


@rigoudyg  In a later PR one can make use of the config file in the query and logger modules. For now I added `log_level` and `log_file` as key words, but they are yet ignored. When your big PR #73 is merged I will make a suggestion for the `logger`. Then I will also give it a try with the consolidation / mapping part again. It may be useful to distinguish the consolidation/mapping from the transformation, but may as well not be.

matthew-mizielinski commented on pyproject.toml:
```diff
@@ -38,6 +39,7 @@ Issues = "https://github.com/CMIP-Data-Request/CMIP7_DReq_Software/issues"
+CMIP7_data_request_api_config = "data_request_api.command_line.config:main"
 
 [tool.pytest.ini_options]
 cache_dir = "tests/.pytest_cache"
```
I think `cache_dir` on line 45 needs a quick update as the `tests` directory no longer exists.

```suggestion
cache_dir = "./.pytest_cache"
```

matthew-mizielinski reviewed (approved):
Subject to the one tweak in the pyproject.toml I think this is good to go.

All tests pass for me.

sol1105 committed:
Update pyproject.toml

Adapting pytest cache_dir

Co-authored-by: Matthew Mizielinski <matthew.mizielinski@metoffice.gov.uk>
